src/aotcompile.cpp

```diff
@@ -592,9 +592,11 @@ void addTargetPasses(legacy::PassManagerBase *PM, TargetMachine *TM)
 void addMachinePasses(legacy::PassManagerBase *PM, TargetMachine *TM, int optlevel)
 {
     // TODO: don't do this on CPUs that natively support Float16
-    PM->add(createDemoteFloat16Pass());
-    if (optlevel > 1)
-        PM->add(createGVNPass());
+    if(TM->getTargetFeatureString().find("fp16") == llvm::StringRef::npos){
```

Comment:
The feature string can includes negatives, like `-fp16`, or possible alias with unrelated features that include `fp16` in their name (albeit unlikely). Anyway, it'd be best to take some extra care here, e.g., https://github.com/JuliaLang/julia/blob/5b4fd5ffac9b9d3aaa407255ea11ad9d43420dd7/src/llvm-cpufeatures.cpp#L62-L73